### PR TITLE
Stats, SNR_in_CC workflow

### PR DESCRIPTION
**Pull request by davhunt**
Given DWI data (mask, bounding box thresholds optional), computes the signal/noise ratio in the corpus callosum

### CODE REVIEW

File: dipy/workflows/stats.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        for data_path, data_bvals_path, data_bvecs_path, out_path in io_it:
+           
+            img = nib.load('{0}'.format(data_path))
+            bvals, bvecs = read_bvals_bvecs('{0}'.format(data_bvals_path), '{0}'.format(data_bvecs_path))
+            gtab = gradient_table(bvals, bvecs)
+            
+            data = img.get_data()
+            affine = img.affine
+
+            if mask == None:
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: you should use `is` here: `if mask is None`

File: dipy/workflows/stats.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+                    SNR = mean_signal[direction]/noise_std
+                SNR_output.append(SNR)
+            
+            data = []
+            data.append({
+                        'data': str(SNR_output[0]) + ' ' + str(SNR_output[1]) + ' ' + str(SNR_output[2]) + ' ' + str(SNR_output[3]),
+                        'directions': 'b0' + ' ' + str(SNR_directions[0]) + ' ' + str(SNR_directions[1]) + ' ' + str(SNR_directions[2])
+                        })
+            
+            with open(os.path.join(out_dir,out_file), 'w') as myfile:
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: pep8: missing whitespace around operator

File: dipy/workflows/stats.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            axis_Y = np.argmin(np.sum((gtab.bvecs-np.array([0, 1, 0])) **2, axis=-1))
+            axis_Z = np.argmin(np.sum((gtab.bvecs-np.array([0, 0, 1])) **2, axis=-1))
+
+            SNR_output = []
+            SNR_directions = []
+            for direction in ['b0', axis_X, axis_Y, axis_Z]:
+                if direction == 'b0':
+                    SNR = mean_signal[0]/noise_std
+                    logging.info("SNR for the b=0 image is :" + str(SNR))
+                else :
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: Can you remove the space between `else` and `:`

File: dipy/workflows/stats.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            mask_noise_img = nib.Nifti1Image(mask_noise.astype(np.uint8), affine)
+            nib.save(mask_noise_img, 'mask_noise.nii.gz')
+            noise_std = np.std(data[mask_noise, :])
+            logging.info('Noise standard deviation sigma= ' + str(noise_std))
+
+            idx = np.sum(gtab.bvecs, axis=-1) == 0
+            gtab.bvecs[idx] = np.inf
+            axis_X = np.argmin(np.sum((gtab.bvecs-np.array([1, 0, 0])) **2, axis=-1))
+            axis_Y = np.argmin(np.sum((gtab.bvecs-np.array([0, 1, 0])) **2, axis=-1))
+            axis_Z = np.argmin(np.sum((gtab.bvecs-np.array([0, 0, 1])) **2, axis=-1))
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: pep8: missing whitespace around operator. Can you add a space after `**`

File: dipy/workflows/stats.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+
+class SNRinCCFlow(Workflow):
+    
+    @classmethod
+    def get_short_name(cls):
+        return 'snrincc'
+    
+    def run(self, data_file, data_bvals, data_bvecs, mask=None, bbox_threshold=(0.6, 1, 0, 0.1, 0, 0.1), out_dir = '', out_file='product.json'):
+        """
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: It would be good to have a short description.

File: dipy/workflows/stats.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+            cfa_img = nib.Nifti1Image((cfa*255).astype(np.uint8), affine)
+            mask_cc_part_img = nib.Nifti1Image(mask_cc_part.astype(np.uint8), affine)
+            nib.save(mask_cc_part_img, 'cc.nii.gz')
+
+            mean_signal = np.mean(data[mask_cc_part], axis=0)
+            mask_noise = binary_dilation(mask, iterations=10)
+            mask_noise[..., :mask_noise.shape[-1]//2] = 1
+            mask_noise = ~mask_noise
+            mask_noise_img = nib.Nifti1Image(mask_noise.astype(np.uint8), affine)
+            nib.save(mask_noise_img, 'mask_noise.nii.gz')
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: It seems you save 2 images (`cc.nii.gz` and `mask_noise.nii.gz`) so it should be an output.

**Comment by skoudoro** [comment]
@pep8speaks Resume now

**Comment by davhunt** [comment]
Hi @skoudoro sorry for the delay-- I think I fixed all of these issues.

File: dipy/workflows/stats.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+from dipy.core.gradients import gradient_table
+from dipy.segment.mask import median_otsu
+from dipy.reconst.dti import TensorModel
+
+from dipy.segment.mask import segment_from_cfa
+from dipy.segment.mask import bounding_box
+
+from dipy.workflows.workflow import Workflow
+
+class SNRinCCFlow(Workflow):
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: pep8: expected 2 blank lines

File: dipy/workflows/stats.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    
+    @classmethod
+    def get_short_name(cls):
+        return 'snrincc'
+    
+    def run(self, data_file, data_bvals, data_bvecs, mask=None, bbox_threshold=(0.6, 1, 0, 0.1, 0, 0.1), out_dir='', out_file='product.json', out_mask_cc='cc.nii.gz', out_mask_noise='mask_noise.nii.gz'):
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: pep8: line too long (203 > 79 characters)

File: dipy/workflows/stats.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: pep8: no space before docstring

File: dipy/workflows/stats.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+            Path to the dwi.nii.gz file. This path may contain wildcards to
+            process multiple inputs at once.
+        data_bvals : string
+            Path of bvals.
+        data_bvecs : string
+            Path of bvecs.
+        mask : string, optional
+            Path of mask if desired. (default None)
+        bbox_threshold : string, optional
+            Threshold for bounding box, values separated with commas for ex. [0.6,1,0,0.1,0,0.1]. (default (0.6, 1, 0, 0.1, 0, 0.1))
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: pep8: line too long (132> 79 characters)

File: dipy/workflows/stats.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+        out_dir : string, optional
+            Where the resulting file will be saved. (default '')
+        out_file : string, optional
+            Name of the result file to be saved. (default 'product.json')
+        out_mask_cc : string, optional
+            Name of the CC mask volume to be saved (default 'cc.nii.gz')
+        out_mask_noise : string, optional
+            Name of the mask noise volume to be saved (default 'mask_noise.nii.gz')
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: pep8: line too long (83> 79 characters)

File: dipy/workflows/stats.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+        """
+        
+        if not isinstance(bbox_threshold, tuple):
+            b = bbox_threshold.replace("[","")
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: pep8: missing white space after ','

File: dipy/workflows/stats.py [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+            b = b.replace("]","")
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: pep8: missing white space after ','

File: dipy/workflows/stats.py [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+            b = b.replace("(","")
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: pep8: missing white space after ','

File: dipy/workflows/stats.py [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+            b = b.replace(")","")
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: pep8: missing white space after ','

File: dipy/workflows/stats.py [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+            b = b.replace(" ","")
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: pep8: missing white space after ','

File: dipy/workflows/stats.py [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+            b = b.split(",")
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: pep8: missing white space after ','

File: dipy/workflows/stats.py [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+            for i in range(len(b)):
+                b[i] = float(b[i])
+            bbox_threshold = tuple(b)
+
+        io_it = self.get_io_iterator()
+        
+        for data_path, data_bvals_path, data_bvecs_path, out_path, cc_mask_path, mask_noise_path in io_it:
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: pep8: line too long (106> 79 characters)

File: dipy/workflows/stats.py [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+            img = nib.load('{0}'.format(data_path))
+            bvals, bvecs = read_bvals_bvecs('{0}'.format(data_bvals_path), '{0}'.format(data_bvecs_path))
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: pep8: line too long (105> 79 characters)

File: dipy/workflows/stats.py [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+            gtab = gradient_table(bvals, bvecs)
+
```
Comment (skoudoro) [inline dipy/workflows/stats.py]: pep8: blank line contains white space

**Comment by davhunt** [comment]
Ah okay. I think now they're all fixed.

**Comment by skoudoro** [comment]
Thank you @davhunt.

I will merge this PR at the end of the day if there is no objection

**Comment by davhunt** [comment]
Thank you!

**Pull request merged**